In [1]:
# Data handling
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Stats
from scipy import stats


In [2]:
df = pd.read_csv('../data/MachineLearningRating_v3.txt',
    sep='|',
    encoding='utf-8',
    engine='python'
)

## Data Understanding

In [3]:
df.head()          # preview first 5 rows
df.info()          # check column types and missing values
df.describe()      # summary statistics for numeric columns


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000098 entries, 0 to 1000097
Data columns (total 52 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   UnderwrittenCoverID       1000098 non-null  int64  
 1   PolicyID                  1000098 non-null  int64  
 2   TransactionMonth          1000098 non-null  object 
 3   IsVATRegistered           1000098 non-null  bool   
 4   Citizenship               1000098 non-null  object 
 5   LegalType                 1000098 non-null  object 
 6   Title                     1000098 non-null  object 
 7   Language                  1000098 non-null  object 
 8   Bank                      854137 non-null   object 
 9   AccountType               959866 non-null   object 
 10  MaritalStatus             991839 non-null   object 
 11  Gender                    990562 non-null   object 
 12  Country                   1000098 non-null  object 
 13  Province                  1

UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  9.995460e+05   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.487770e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360381e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.005692e+07   
50%           9.408300e+04  7.071000e+03  2.000000e+03  6.005842e+07   
75%           1.391900e+05  1.107700e+04  4.180000e+03  6.005842e+07   
max           3.011750e+05  2.324600e+04  9.870000e+03  6.506535e+07   

       RegistrationYear      Cylinders  cubiccapacity      kilowatts  \
count      1.000098e+06  999546.000000  999546.000000  999546.000000   
mean       2.010225e+03       4.046642    2466.743258      97.207919   
std        3.261391e+00       0.294020     442.800640      19.393256   
min        1.987000e+03       0.000000       0.000000       0.000000   
25%        2.008000e+03       4.000000    2237.000000      75.000000   
50%        2.011000e+03       4.000000    2694.000000     111.000000   
75%        2.013000e+03       4.000000    2694.000000     111.000000   
max        2.015000e+03      10.000000   12880.000000     309.000000   

       NumberOfDoors  CustomValueEstimate  NumberOfVehiclesInFleet  \
count  999546.000000         2.204560e+05                      0.0   
mean        4.019250         2.255311e+05                      NaN   
std         0.468314         5.645157e+05                      NaN   
min         0.000000         2.000000e+04                      NaN   
25%         4.000000         1.350000e+05                      NaN   
50%         4.000000         2.200000e+05                      NaN   
75%         4.000000         2.800000e+05                      NaN   
max         6.000000         2.655000e+07                      NaN   

         SumInsured  CalculatedPremiumPerTerm  TotalPremium   TotalClaims  
count  1.000098e+06              1.000098e+06  1.000098e+06  1.000098e+06  
mean   6.041727e+05              1.178757e+02  6.190550e+01  6.486119e+01  
std    1.508332e+06              3.997017e+02  2.302845e+02  2.384075e+03  
min    1.000000e-02              0.000000e+00 -7.825768e+02 -1.200241e+04  
25%    5.000000e+03              3.224800e+00  0.000000e+00  0.000000e+00  
50%    7.500000e+03              8.436900e+00  2.178333e+00  0.000000e+00  
75%    2.500000e+05              9.000000e+01  2.192982e+01  0.000000e+00  
max    1.263620e+07              7.442217e+04  6.528260e+04  3.930921e+05

Based on this data snippet, here's what we can understand:

##### 🔍 **Key Observations from Insurance Data:**

##### **Data Structure:**
- **52 columns** including policy IDs, vehicle details, coverage info, and financial metrics
- **Mixed data types**: numeric, categorical, and possibly missing values (NaN)

##### **Vehicle & Policy Details:**
- **Vehicle specifications**: Registration year, cylinders, cubic capacity, kilowatts, doors
- **Coverage details**: Sum insured, premiums, claims data
- **Policy identifiers**: UnderwrittenCoverID, PolicyID, PostalCode

##### **Financial Metrics:**
- **Premium calculations**: CalculatedPremiumPerTerm, TotalPremium
- **Risk assessment**: Multiple premium/claim amount columns
- **Fleet data**: NumberOfVehiclesinFleet indicates commercial policies

##### **Data Quality Issues:**
- **Missing values**: NaN present in multiple columns
- **Formatting problems**: Scientific notation mixed with regular numbers
- **Inconsistent data**: Some columns showing extreme values or placeholders

##### **Potential Analysis Areas:**
1. **Risk modeling** using vehicle specs vs claim amounts
2. **Pricing optimization** based on calculated vs actual premiums
3. **Fraud detection** from claim patterns
4. **Customer segmentation** by postal code/vehicle type

**Next steps**: Clean the data, handle missing values, convert formats, then explore relationships between vehicle characteristics and insurance claims.

### Check Missing Values


In [4]:
missing = df.isnull().sum().sort_values(ascending=False)
missing_percent = (missing / len(df)) * 100
missing_table = pd.concat([missing, missing_percent], axis=1, keys=['Missing Values', '% of Total'])
print(missing_table)


                          Missing Values  % of Total
NumberOfVehiclesInFleet          1000098  100.000000
CrossBorder                       999400   99.930207
CustomValueEstimate               779642   77.956560
Rebuilt                           641901   64.183810
Converted                         641901   64.183810
WrittenOff                        641901   64.183810
NewVehicle                        153295   15.327998
Bank                              145961   14.594670
AccountType                        40232    4.022806
Gender                              9536    0.953507
MaritalStatus                       8259    0.825819
VehicleType                          552    0.055195
make                                 552    0.055195
mmcode                               552    0.055195
Model                                552    0.055195
Cylinders                            552    0.055195
bodytype                             552    0.055195
kilowatts                            552    0.

###### 🔍 **Missing Value Analysis**

###### **Critical Issues:**
- **100% missing**: `NumberOfVehiclesInFleet` → Column is **useless**
- **99.9% missing**: `CrossBorder` → Nearly useless
- **64-78% missing**: Key vehicle history fields (`Rebuilt`, `Converted`, `WrittenOff`, `CustomValueEstimate`)

###### **Data Quality:**
- **Perfect columns**: 30+ columns have **0% missing** (policy, premium, claim data)
- **Vehicle specs problematic**: 0.06% missing in make/model/engine details
- **Demographics issues**: Gender/MaritalStatus ~1% missing

###### **Action Required:**
1. **Drop**: 100% missing columns
2. **Impute/Careful handling**: High-missing columns (>60%)
3. **Minimal impact**: Low-missing columns (<1%)

**Bottom line**: Core insurance data (premiums, claims, policy info) is **complete**, but vehicle/demographic data needs cleaning.

### Check column types

In [5]:
df['TransactionMonth'] = pd.to_datetime(df['TransactionMonth'])
df['VehicleIntroDate'] = pd.to_datetime(df['VehicleIntroDate'])


C:\Users\zkidanemariam\AppData\Local\Temp\ipykernel_15888\3243247046.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['VehicleIntroDate'] = pd.to_datetime(df['VehicleIntroDate'])
